plugin.go

```diff
@@ -0,0 +1,111 @@
+package containerd
+
+import (
+	"fmt"
+	"path/filepath"
+	"plugin"
+	"runtime"
+	"sync"
+
+	"github.com/docker/containerd/content"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+type PluginType int
+
+const (
+	RuntimePlugin PluginType = iota + 1
+	GRPCPlugin
+)
+
+type Registration struct {
+	Type   PluginType
+	Config interface{}
+	Init   func(*InitContext) (interface{}, error)
+}
+
+type InitContext struct {
+	Root     string
+	State    string
+	Runtimes map[string]Runtime
+	Store    *content.Store
+	Config   interface{}
+	Context  context.Context
+}
+
+type Service interface {
+	Register(*grpc.Server) error
+}
+
+var register = struct {
+	sync.Mutex
+	r map[string]*Registration
+}{
+	r: make(map[string]*Registration),
+}
+
+// Load loads all plugins at the provided that into containerd
+func Load(path string) (err error) {
+	defer func() {
+		if v := recover(); v != nil {
+			rerr, ok := v.(error)
+			if !ok {
+				panic(v)
```

Comment:
If the goal of this is to protect plugins from causing a panic on load then why propagate when not an error? If the panic was initiated without an error such as `panic("something is wrong")`, then it wouldn't get caught. Maybe either declare an error type which has an `interface{}` sub-value or set the error like `fmt.Errorf('plugin panic: %v", v)`.